### PR TITLE
Improve error messages for ValidateObject method.

### LINKED ISSUE #39110: Clarify container.name validation failures
create this deamonset
```yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: daemon-test
  labels:
    app: daemon-test
spec:
  selector:
    matchLabels:
      app: daemon-test
  template:
    metadata:
      labels:
        app: daemon-test
    spec:
      containers:
      - image: busybox
        command:
        - sleep
        - "36000"
        resources:
          limits:
            cpu: 200m
            memory: 100Mi
          requests:
            cpu: 20m
            memory: 20Mi
      restartPolicy: Always
```

with 
```console
$ kubectl create -f daemonset.yaml
error: error validating "/tmp/daemon.yaml": error validating data: field name: is required; if you choose to ignore these errors, turn validation off with --validate=false
```

Which is a confusing error. 
The code to change is: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L1717
It can mention that path of the parent object missing the name.

### PR DESCRIPTION
**Pull request by xingzhou**
Improved error messages for #ValidateObject method
in pkg/api/validation/schema.go.

Fixed #39110

### CODE REVIEW

**Comment by lavalamp** [comment]
@k8s-bot ok to test

Sorry for delay, I was OOO. I think @mbohlool is best suited to review this.

File: pkg/api/validation/schema.go [hunk 1]
```diff
@@ -253,7 +253,7 @@ func (s *SwaggerSchema) ValidateObject(obj interface{}, fieldName, typeName stri
 	if !ok && s.delegate != nil {
 		fields, mapOk := obj.(map[string]interface{})
 		if !mapOk {
-			return append(allErrs, fmt.Errorf("field %s: expected object of type map[string]interface{}, but the actual type is %T", fieldName, obj))
+			return append(allErrs, fmt.Errorf("expected object '%s' of type 'map[string]interface{}', but the actual type is '%T'", fieldName, obj))
```
Comment (mbohlool) [inline pkg/api/validation/schema.go]: I tried to not be picky, but not sure what value those single quote characters adds. What appears between them does not have spaces and can be easily identified in the error message.

File: pkg/api/validation/schema.go [hunk 2]
```diff
@@ -273,15 +273,15 @@ func (s *SwaggerSchema) ValidateObject(obj interface{}, fieldName, typeName stri
...
+		return append(allErrs, fmt.Errorf("expected object '%s' of type 'map[string]interface{}', but the actual type is '%T'", fieldName, obj))
 	}
 	if len(fieldName) > 0 {
 		fieldName = fieldName + "."
 	}
 	// handle required fields
 	for _, requiredKey := range model.Required {
 		if _, ok := fields[requiredKey]; !ok {
-			allErrs = append(allErrs, fmt.Errorf("field %s: is required", requiredKey))
+			allErrs = append(allErrs, fmt.Errorf("field '%s%s' is required, but not provided", fieldName, requiredKey))
```
Comment (mbohlool) [inline pkg/api/validation/schema.go]: A validation error "is required" implies "not provided".

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (mbohlool) [inline pkg/api/validation/schema.go]: I tried to not be picky, but not sure what value those single quote characters adds. What appears between them does not have spaces and can be easily identified in the error message. ditto for the rest of the change.

**Review by mbohlool — CHANGES_REQUESTED** [review]

**Comment by xingzhou** [comment]
thanks @mbohlool for the review, have updated the code according to the comments, please take a look

**Comment by xingzhou** [comment]
rebased

**Review by mbohlool — APPROVED** [review]

**Comment by xingzhou** [comment]
@erictune, would you please help to review and approve this patch? thanks

**Review by mbohlool — APPROVED** [review]

**Comment by xingzhou** [comment]
@k8s-bot bazel test this

**Comment by xingzhou** [comment]
@k8s-bot bazel test this

**Pull request merged**